File: atm cajero sockets/cajero-cliente/src/atm/client/session.py
"""
Equivalente a lenin.client.Session del profesor.

Java usa ObjectOutputStream/ObjectInputStream (writeObject/readObject).
Python no tiene eso "de fábrica" para sockets, así que armamos un framing
simple sobre los file objects del socket (socket.makefile): cada objeto
se serializa a JSON y se antecede por su longitud en 4 bytes (big-endian),
para que read() sepa exactamente cuántos bytes leer sin adivinar dónde
termina un mensaje y empieza el siguiente.
"""

import json
import socket
import struct


class Session:

    def __init__(self, sock: socket.socket):
        self.socket = sock
        try:
            self._wfile = sock.makefile("wb")
            self._rfile = sock.makefile("rb")
        except Exception as e:
            print(f"Error creando la sesión: {e}")
            self._wfile = None
            self._rfile = None
            self.socket = None

    def read(self):
        try:
            header = self._rfile.read(4)
            if not header or len(header) < 4:
                return None
            (longitud,) = struct.unpack("!I", header)
            payload = self._rfile.read(longitud)
            if len(payload) < longitud:
                return None
            return json.loads(payload.decode("utf-8"))
        except (OSError, AttributeError, ValueError, json.JSONDecodeError):
            return None

    def write(self, data) -> bool:
        try:
            payload = json.dumps(data).encode("utf-8")
            self._wfile.write(struct.pack("!I", len(payload)))
            self._wfile.write(payload)
            self._wfile.flush()
            return True
        except (OSError, AttributeError, TypeError, ValueError) as e:
            print(f"Error escribiendo en la sesión: {e}")
            return False

    def close(self) -> bool:
        try:
            if self._wfile:
                self._wfile.close()
            if self._rfile:
                self._rfile.close()
            if self.socket:
                self.socket.close()
            return True
        except OSError as e:
            print(f"Error cerrando la sesión: {e}")
            return False

File: atm cajero sockets/cajero-cliente-go/src/atm/client/session.go
// Equivalente a atm/client/session.py. Java usa
// ObjectOutputStream/ObjectInputStream (writeObject/readObject); ni Python
// ni Go tienen eso "de fábrica" para sockets, así que armamos el mismo
// framing manual sobre la conexión TCP:
//
//	[ 4 bytes big-endian: cuántos bytes vienen ][ N bytes: el JSON en sí ]
//
// para que Read() sepa exactamente cuántos bytes leer sin adivinar dónde
// termina un mensaje y empieza el siguiente.
package client

import (
	"encoding/binary"
	"encoding/json"
	"net"
)

type Session struct {
	conn net.Conn
}

func NewSession(conn net.Conn) *Session {
	return &Session{conn: conn}
}

// recvExact lee exactamente n bytes, haciendo loop si hace falta: recv()
// no siempre entrega todos los bytes pedidos de una sola vez.
func (s *Session) recvExact(n int) ([]byte, error) {
	buf := make([]byte, n)
	leidos := 0
	for leidos < n {
		m, err := s.conn.Read(buf[leidos:])
		if err != nil {
			return nil, err
		}
		leidos += m
	}
	return buf, nil
}

func (s *Session) Read() (interface{}, error) {
	header, err := s.recvExact(4)
	if err != nil {
		return nil, err
	}
	longitud := binary.BigEndian.Uint32(header)
	payload, err := s.recvExact(int(longitud))
	if err != nil {
		return nil, err
	}
	var dato interface{}
	if err := json.Unmarshal(payload, &dato); err != nil {
		return nil, err
	}
	return dato, nil
}

func (s *Session) Write(dato interface{}) bool {
	payload, err := json.Marshal(dato)
	if err != nil {
		return false
	}
	header := make([]byte, 4)
	binary.BigEndian.PutUint32(header, uint32(len(payload)))
	if _, err := s.conn.Write(append(header, payload...)); err != nil {
		return false
	}
	return true
}

func (s *Session) Close() bool {
	return s.conn.Close() == nil
}
